homeassistant/components/zha/config_flow.py

```diff
@@ -250,55 +634,107 @@ async def async_step_hardware(self, data=None):
-    for radio in RadioType:
-        dev_config = radio.controller.SCHEMA_DEVICE({CONF_DEVICE_PATH: dev_path})
-        probe_result = await radio.controller.probe(dev_config)
-        if probe_result:
-            if isinstance(probe_result, dict):
-                return {CONF_RADIO_TYPE: radio.name, CONF_DEVICE: probe_result}
-            return {CONF_RADIO_TYPE: radio.name, CONF_DEVICE: dev_config}
+class ZhaOptionsFlowHandler(ZhaFlowMixin, config_entries.OptionsFlow):
+    """Handle an options flow."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initialize options flow."""
+        super().__init__()
+        self.config_entry = config_entry
+
+        self._device_path = config_entry.data[CONF_DEVICE][CONF_DEVICE_PATH]
+        self._device_settings = config_entry.data[CONF_DEVICE]
+        self._radio_type = RadioType[config_entry.data[CONF_RADIO_TYPE]]
+        self._title = config_entry.title
+
+        self._must_reload_entry = False
+
+    async def async_step_init(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Launch the options flow."""
+        if user_input is not None:
+            try:
+                await self.hass.config_entries.async_unload(self.config_entry.entry_id)
+            except config_entries.OperationNotAllowed:
+                # ZHA is not running
+                pass
+
+            self._must_reload_entry = True
+
+            return await self.async_step_choose_serial_port()
+
+        return self.async_show_form(step_id="init")
+
+    async def _async_create_radio_entity(self):
+        """Re-implementation of the base flow's final step to update the config."""
+        device_settings = self._device_settings.copy()
+        device_settings[CONF_DEVICE_PATH] = await self.hass.async_add_executor_job(
+            usb.get_serial_by_id, self._device_path
+        )
+
+        # Avoid creating both `.options` and `.data` by directly writing `data` here
+        self.hass.config_entries.async_update_entry(
+            entry=self.config_entry,
+            data={
+                CONF_DEVICE: device_settings,
+                CONF_RADIO_TYPE: self._radio_type.name,
+            },
+            options=self.config_entry.options,
+        )
+
+        # Reload ZHA after we finish
+        self._must_reload_entry = False
+        await self.hass.config_entries.async_setup(self.config_entry.entry_id)
+
+        # Intentionally do not set `data` to avoid creating `options`, we set it above
+        return self.async_create_entry(title=self._title, data={})
+
+    def async_remove(self):
+        """Maybe reload ZHA if the flow is aborted."""
+        if self._must_reload_entry:
+            setup = self.hass.config_entries.async_setup(self.config_entry.entry_id)
 
-    return None
+            try:
+                self.hass.async_create_task(setup)
+            except RuntimeError:
```

Comment:
Why do we need to handle this? Doesn't the core take care of finishing tasks before closing the loop?